src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -79,13 +92,246 @@ struct C_EncryptedObjectReadRequest : public Context {
           r += extent.length;
         }
       }
-      onfinish->complete(r);
+
+      if (r == -ENOENT && read_parent_sender != nullptr) {
+        read_parent_sender();
+      } else {
+        on_finish->complete(r);
+      }
+    }
+};
+
+template <typename I>
+struct C_UnalignedObjectReadRequest : public Context {
+
+    CephContext* cct;
+    io::ReadExtents* extents;
+    io::ReadExtents aligned_extents;
+    Context* on_finish;
+    io::ObjectDispatchSpec* req;
+
+    C_UnalignedObjectReadRequest(
+            I* image_ctx, ceph::ref_t<CryptoInterface> crypto,
+            uint64_t object_no, io::ReadExtents* extents, IOContext io_context,
+            int op_flags, int read_flags, const ZTracer::Trace &parent_trace,
+            uint64_t* version, int* object_dispatch_flags,
+            Context* on_dispatched) : cct(image_ctx->cct),
+                                      extents(extents),
+                                      on_finish(on_dispatched) {
+      crypto->align_extents(*extents, &aligned_extents);
+
+      // send the aligned read back to get decrypted
+      req = io::ObjectDispatchSpec::create_read(
+              image_ctx, io::OBJECT_DISPATCH_LAYER_NONE, object_no,
+              &aligned_extents, io_context, op_flags, read_flags, parent_trace,
+              version, this);
+    }
+
+    void send() {
+      req->send();
+    }
+
+    void finish(int r) override {
+      ldout(cct, 20) << "r=" << r << dendl;
+      if (r >= 0) {
+        r = 0;
+        for (uint64_t i = 0; i < extents->size(); ++i) {
+          auto& extent = (*extents)[i];
+          auto& aligned_extent = aligned_extents[i];
+          extent.bl.substr_of(
+                  aligned_extent.bl,
+                  extent.offset - aligned_extent.offset,
+                  extent.length);
+          r += extent.length;
+        }
+      }
+      on_finish->complete(r);
+    }
+};
+
+template <typename I>
+struct C_UnalignedObjectWriteRequest : public Context {
+
+    CephContext* cct;
+    uint64_t object_off;
+    ceph::bufferlist* aligned_data;
```

Comment:
Nit: split up into `data`, `aligned_data`, and `cmp_data` -- where only `aligned_data` changes and `data` and `cmd_data` are kept immutable in case of restart

Reply:
How do we then capture `data` in the lambda expression?
We need to capture it without copying it, to match the performance of the current solution
We cannot capture it by reference since the C_UnalignedObjectWriteRequest struct will already be deleted once this lambda expression executes.

Reply:
Instead of having the `read` call invoke `complete`, you have create a `handle_read(int)` method so that your object doesn't get deleted while you are still using it. Then you can avoid the majority of captures since you can just re-use `this` as a capture.